src/components/Table/utils/TableState.ts

```diff
@@ -218,25 +218,61 @@ export class TableState {
   setColumns(columns: GridColumnWithId<any>[], visibleColumnsStorageKey: string | undefined): void {
     const isInitial = !this.columns || this.columns.length === 0;
     if (columns !== this.columns) {
-      this.columns = columns;
       this.visibleColumnsStorageKey = visibleColumnsStorageKey ?? camelCase(columns.map((c) => c.id).join());
       this.visibleColumns.replace(readOrSetLocalVisibleColumnState(columns, this.visibleColumnsStorageKey));
-      const expandedColumnIds = columns.filter((c) => c.initExpanded).map((c) => c.id);
-      if (isInitial) this.expandedColumns.replace(expandedColumnIds);
+      // list of local storage columns
+      const localStorageColumns = this.persistCollapse ? readExpandedColumnsStorage(this.persistCollapse) : [];
+      // expandedColumnIds initialized as localStorageColumns
+      const existingExpandedColumnIds = isInitial ? localStorageColumns : this.expandedColumnIds;
+
+      // list of all existing columns
+      const existingColumnIds = this.columns.map((c) => c.id);
+
+      // filter existing columns that are initExpanded to see which are new
```

Comment:
**nit:** The comment here is misleading. We are filtering new columns, not the existing. Maybe something like:
```suggestion
      // filter to find new columns that should be initially expanded
```